[Bug]: test_image_comparison_expect_rms nondeterministic failure

### Bug summary

This seems a rather rare occurrence, but I've gotten the test to fail like this, on a x86_64 processor:
```
    def test_image_comparison_expect_rms(im1, im2, tol, expect_rms):
        """
        Compare two images, expecting a particular RMS error.
    
        im1 and im2 are filenames relative to the baseline_dir directory.
    
        tol is the tolerance to pass to compare_images.
    
        expect_rms is the expected RMS value, or None. If None, the test will
        succeed if compare_images succeeds. Otherwise, the test will succeed if
        compare_images fails and returns an RMS error almost equal to this value.
        """
        baseline_dir, result_dir = map(Path, _image_directories(lambda: "dummy"))
        # Copy both "baseline" and "test" image to result_dir, so that 1)
        # compare_images writes the diff to result_dir, rather than to the source
        # tree and 2) the baseline image doesn't appear missing to triage_tests.py.
        result_im1 = make_test_filename(result_dir / im1, "expected")
        shutil.copyfile(baseline_dir / im1, result_im1)
        result_im2 = result_dir / im1
        shutil.copyfile(baseline_dir / im2, result_im2)
        results = compare_images(
            result_im1, result_im2, tol=tol, in_decorator=True)
    
        if expect_rms is None:
            assert results is None
        else:
            assert results is not None
>           assert results['rms'] == approx(expect_rms, abs=1e-4)
E           assert 90.15611460128481 == 6.50646 ± 1.0e-04
E            +  where 6.50646 ± 1.0e-04 = approx(6.50646, abs=0.0001)
```

### Code for reproduction

```python
Run the test suite via pytest. In my case, the test suite is run in parallel with xdist, in case this matters. As stated, this is nondeterministic, so it may be difficult to reproduce.
```


### Actual outcome

All tests should pass.

### Expected outcome

The `test_image_comparison_expect_rms` fails  as shown in the bug summary.

### Additional information

Attached complete build log: 
[ffvkxha5p34br976578fm3vz8hppis-python-matplotlib-3.5.1.drv.txt](https://github.com/matplotlib/matplotlib/files/8640333/ffvkxha5p34br976578fm3vz8hppis-python-matplotlib-3.5.1.drv.txt)


### Operating system

GNU Guix

### Matplotlib Version

3.5.1 (the test is unchanged in 3.5.2)

### Matplotlib Backend

_No response_

### Python version

3.9.9

### Jupyter version

_No response_

### Installation

from source (.tar.gz)

--- a/lib/matplotlib/tests/test_compare_images.py
+++ b/lib/matplotlib/tests/test_compare_images.py
@@ -60,12 +60,12 @@ def test_image_comparison_expect_rms(im1, im2, tol, expect_rms, tmp_path,
     # Copy both "baseline" and "test" image to result_dir, so that 1)
     # compare_images writes the diff to result_dir, rather than to the source
     # tree and 2) the baseline image doesn't appear missing to triage_tests.py.
-    result_im1 = make_test_filename(result_dir / im1, "expected")
-    shutil.copyfile(baseline_dir / im1, result_im1)
+    # result_im1 = make_test_filename(result_dir / im1, "expected")
+    # shutil.copyfile(baseline_dir / im1, result_im1)
     result_im2 = result_dir / im1
     shutil.copyfile(baseline_dir / im2, result_im2)
     results = compare_images(
-        result_im1, result_im2, tol=tol, in_decorator=True)
+        baseline_dir / im1, result_im2, tol=tol, in_decorator=True)
 
     if expect_rms is None:
         assert results is None

--- a/lib/matplotlib/tests/test_compare_images.py
+++ b/lib/matplotlib/tests/test_compare_images.py
@@ -55,6 +55,8 @@ def test_image_comparison_expect_rms(im1, im2, tol, expect_rms, tmp_path,
     succeed if compare_images succeeds. Otherwise, the test will succeed if
     compare_images fails and returns an RMS error almost equal to this value.
     """
+    # Change the working directory using monkeypatch to use a temporary
+    # test specific directory
     monkeypatch.chdir(tmp_path)
     baseline_dir, result_dir = map(Path, _image_directories(lambda: "dummy"))
     # Copy "test" image to result_dir, so that compare_images writes

--- a/lib/matplotlib/tests/test_compare_images.py
+++ b/lib/matplotlib/tests/test_compare_images.py
@@ -57,11 +57,8 @@ def test_image_comparison_expect_rms(im1, im2, tol, expect_rms, tmp_path,
     """
     monkeypatch.chdir(tmp_path)
     baseline_dir, result_dir = map(Path, _image_directories(lambda: "dummy"))
-    # Copy both "baseline" and "test" image to result_dir, so that 1)
-    # compare_images writes the diff to result_dir, rather than to the source
-    # tree and 2) the baseline image doesn't appear missing to triage_tests.py.
-    # result_im1 = make_test_filename(result_dir / im1, "expected")
-    # shutil.copyfile(baseline_dir / im1, result_im1)
+    # Copy "test" image to result_dir, so that compare_images writes
+    # the diff to result_dir, rather than to the source tree
     result_im2 = result_dir / im1
     shutil.copyfile(baseline_dir / im2, result_im2)
     results = compare_images(

--- a/lib/matplotlib/tests/test_compare_images.py
+++ b/lib/matplotlib/tests/test_compare_images.py
@@ -4,7 +4,7 @@ import shutil
 import pytest
 from pytest import approx
 
-from matplotlib.testing.compare import compare_images, make_test_filename
+from matplotlib.testing.compare import compare_images
 from matplotlib.testing.decorators import _image_directories
 
 

--- a/lib/matplotlib/tests/test_compare_images.py
+++ b/lib/matplotlib/tests/test_compare_images.py
@@ -1,6 +1,6 @@
 from pathlib import Path
 import shutil
-
+import uuid
 import pytest
 from pytest import approx
 
@@ -58,7 +58,8 @@ def test_image_comparison_expect_rms(im1, im2, tol, expect_rms):
     # Copy both "baseline" and "test" image to result_dir, so that 1)
     # compare_images writes the diff to result_dir, rather than to the source
     # tree and 2) the baseline image doesn't appear missing to triage_tests.py.
-    result_im1 = make_test_filename(result_dir / im1, "expected")
+    uid = str(uuid.uuid4())
+    result_im1 = make_test_filename(result_dir / (uid + im1), "expected")
     shutil.copyfile(baseline_dir / im1, result_im1)
     result_im2 = result_dir / im1
     shutil.copyfile(baseline_dir / im2, result_im2)

--- a/lib/matplotlib/tests/test_compare_images.py
+++ b/lib/matplotlib/tests/test_compare_images.py
@@ -1,6 +1,6 @@
 from pathlib import Path
 import shutil
-import uuid
+
 import pytest
 from pytest import approx
 
@@ -58,8 +58,7 @@ def test_image_comparison_expect_rms(im1, im2, tol, expect_rms):
     # Copy both "baseline" and "test" image to result_dir, so that 1)
     # compare_images writes the diff to result_dir, rather than to the source
     # tree and 2) the baseline image doesn't appear missing to triage_tests.py.
-    uid = str(uuid.uuid4())
-    result_im1 = make_test_filename(result_dir / (uid + im1), "expected")
+    result_im1 = make_test_filename(result_dir / im1, "expected")
     shutil.copyfile(baseline_dir / im1, result_im1)
     result_im2 = result_dir / im1
     shutil.copyfile(baseline_dir / im2, result_im2)

--- a/lib/matplotlib/tests/test_compare_images.py
+++ b/lib/matplotlib/tests/test_compare_images.py
@@ -42,7 +42,8 @@ from matplotlib.testing.decorators import _image_directories
         # Now test the reverse comparison.
         ('all128.png', 'all127.png', 0, 1),
     ])
-def test_image_comparison_expect_rms(im1, im2, tol, expect_rms):
+def test_image_comparison_expect_rms(im1, im2, tol, expect_rms, tmp_path,
+                                                            monkeypatch):
     """
     Compare two images, expecting a particular RMS error.
 
@@ -54,6 +55,7 @@ def test_image_comparison_expect_rms(im1, im2, tol, expect_rms):
     succeed if compare_images succeeds. Otherwise, the test will succeed if
     compare_images fails and returns an RMS error almost equal to this value.
     """
+    monkeypatch.chdir(tmp_path)
     baseline_dir, result_dir = map(Path, _image_directories(lambda: "dummy"))
     # Copy both "baseline" and "test" image to result_dir, so that 1)
     # compare_images writes the diff to result_dir, rather than to the source

--- a/lib/matplotlib/tests/test_compare_images.py
+++ b/lib/matplotlib/tests/test_compare_images.py
@@ -43,7 +43,7 @@ from matplotlib.testing.decorators import _image_directories
         ('all128.png', 'all127.png', 0, 1),
     ])
 def test_image_comparison_expect_rms(im1, im2, tol, expect_rms, tmp_path,
-                                                            monkeypatch):
+                                     monkeypatch):
     """
     Compare two images, expecting a particular RMS error.
 

